File: intellij-plugin/src/test/kotlin/com/ymkang/gitcryptexplorer/scenarios/GitCryptSetupScenarioTest.kt
package com.ymkang.gitcryptexplorer.scenarios

import com.ymkang.gitcryptexplorer.core.GitClient
import com.ymkang.gitcryptexplorer.core.GitCryptLocalState
import com.ymkang.gitcryptexplorer.core.GitCryptRepositoryStatus
import com.ymkang.gitcryptexplorer.core.GitCryptSetupRecommendation
import com.ymkang.gitcryptexplorer.core.gitCryptTargetOperationError
import com.ymkang.gitcryptexplorer.core.recommendGitCryptSetup
import java.nio.file.Files
import org.junit.Assert.assertEquals
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

class GitCryptSetupScenarioTest {
    @get:Rule
    val temporaryFolder = TemporaryFolder()

    @Test
    fun emptyFolderCanBeInitializedAndDiscoveredAsGitRepository() {
        val folder = temporaryFolder.newFolder("empty-project").toPath()
        val git = GitClient()

        assertEquals(
            GitCryptSetupRecommendation.INITIALIZE_GIT,
            recommendGitCryptSetup(repositoryExists = false),
        )

        git.initialize(folder)
        val location = git.discover(folder)

        assertEquals(folder.toRealPath(), location.root)
        assertTrue(java.nio.file.Files.isDirectory(location.gitDir))
    }

    @Test
    fun protectedFilesInLockedRepositoryRecommendUnlock() {
        assertEquals(
            GitCryptSetupRecommendation.UNLOCK,
            recommendGitCryptSetup(
                repositoryExists = true,
                protectedFileCount = 1,
                localState = GitCryptLocalState.LOCKED,
            ),
        )
    }

    @Test
    fun repositoryWithoutTargetsOrLocalKeyRecommendsNewGitCryptKey() {
        assertEquals(
            GitCryptSetupRecommendation.INITIALIZE_GIT_CRYPT,
            recommendGitCryptSetup(
                repositoryExists = true,
                protectedFileCount = 0,
                localState = GitCryptLocalState.NOT_INITIALIZED,
            ),
        )
    }

    @Test
    fun encryptionTargetChangesRequireAnInstalledUnlockedKey() {
        assertEquals(
            "git-crypt is locked. Unlock the repository before changing encryption targets.",
            gitCryptTargetOperationError(
                GitCryptRepositoryStatus(
                    available = true,
                    localState = GitCryptLocalState.LOCKED,
                    installedKeyCount = 0,
                ),
            ),
        )
        assertEquals(
            "git-crypt is not initialized locally. Initialize or unlock the repository before changing encryption targets.",
            gitCryptTargetOperationError(
                GitCryptRepositoryStatus(
                    available = true,
                    localState = GitCryptLocalState.NOT_INITIALIZED,
                    installedKeyCount = 0,
                ),
            ),
        )
        assertNull(
            gitCryptTargetOperationError(
                GitCryptRepositoryStatus(
                    available = true,
                    localState = GitCryptLocalState.UNLOCKED,
                    installedKeyCount = 1,
                ),
            ),
        )
    }

    @Test
    fun encryptedIndexContentCanBeReadWithoutChangingTheWorkingTree() {
        val folder = temporaryFolder.newFolder("indexed-ciphertext").toPath()
        val encryptedContents = byteArrayOf(0, 0x47, 0x49, 0x54, 0x43, 0x52, 0x59, 0x50, 0x54, 0, 1)
        val git = GitClient()
        git.initialize(folder)
        Files.write(folder.resolve("secret.py"), encryptedContents)
        runGit(folder, "add", "--", "secret.py")
        val statusBefore = runGit(folder, "status", "--short", "--", "secret.py")

        val blob = git.readIndexedBlob(folder, "secret.py")

        assertEquals("secret.py", blob.path)
        assertArrayEquals(encryptedContents, blob.contents)
        assertEquals(statusBefore, runGit(folder, "status", "--short", "--", "secret.py"))
    }

    private fun runGit(folder: java.nio.file.Path, vararg arguments: String): String {
        val process = ProcessBuilder(listOf("git", "-C", folder.toString()) + arguments).start()
        val output = process.inputStream.bufferedReader().readText()
        check(process.waitFor() == 0) { process.errorStream.bufferedReader().readText() }
        return output
    }
}

File: test/scenarios/gitCryptSetupScenario.test.ts
import assert from 'node:assert/strict';
import { execFile } from 'node:child_process';
import { mkdtemp, realpath, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import * as path from 'node:path';
import { promisify } from 'node:util';
import { GitClient } from '../../src/git/gitClient';
import {
  gitCryptTargetOperationError,
  recommendGitCryptSetup,
} from '../../src/gitCrypt/setupRecommendation';

import { test, TestContext } from 'node:test';

const execFileAsync = promisify(execFile);

test('SC-001: empty folder can be initialized and discovered as a Git repository', async (t) => {
  const folder = await createEmptyFolder(t);
  const git = new GitClient();

  assert.equal(recommendGitCryptSetup(false), 'initialize-git');

  await git.initialize(folder);
  const location = await git.discover(folder);

  assert.equal(location.root, await realpath(folder));
  assert.equal((await execFileAsync('git', ['rev-parse', '--is-inside-work-tree'], { cwd: folder })).stdout.trim(), 'true');
});

test('SC-002: protected files in a locked repository recommend unlock', () => {
  assert.equal(recommendGitCryptSetup(true, 1, 'locked'), 'unlock');
});

test('SC-003: repository without targets or a local key recommends a new git-crypt key', () => {
  assert.equal(recommendGitCryptSetup(true, 0, 'not-initialized'), 'initialize-git-crypt');
});

test('SC-004: encryption target changes require an installed unlocked key', () => {
  assert.equal(
    gitCryptTargetOperationError({
      available: true,
      localState: 'locked',
      installedKeyCount: 0,
    }),
    'git-crypt is locked. Unlock the repository before changing encryption targets.',
  );
  assert.equal(
    gitCryptTargetOperationError({
      available: true,
      localState: 'unlocked',
      installedKeyCount: 1,
    }),
    undefined,
  );
  assert.equal(
    gitCryptTargetOperationError({
      available: true,
      localState: 'not-initialized',
      installedKeyCount: 0,
    }),
    'git-crypt is not initialized locally. Initialize or unlock the repository before changing encryption targets.',
  );
});

test('SC-005: encrypted index content can be read without changing the working tree', async (t) => {
  const folder = await createEmptyFolder(t);
  const encryptedContents = Buffer.from([0x00, 0x47, 0x49, 0x54, 0x43, 0x52, 0x59, 0x50, 0x54, 0x00, 0x01]);
  const git = new GitClient();
  await git.initialize(folder);
  await writeFile(path.join(folder, 'secret.py'), encryptedContents);
  await execFileAsync('git', ['add', '--', 'secret.py'], { cwd: folder });
  const statusBefore = (await execFileAsync('git', ['status', '--short', '--', 'secret.py'], { cwd: folder })).stdout;

  const blob = await git.readIndexedBlob(folder, 'secret.py');

  assert.equal(blob.path, 'secret.py');
  assert.equal(blob.contents.equals(encryptedContents), true);
  assert.equal((await execFileAsync('git', ['status', '--short', '--', 'secret.py'], { cwd: folder })).stdout, statusBefore);
});

async function createEmptyFolder(t: TestContext): Promise<string> {
  const folder = await mkdtemp(path.join(tmpdir(), 'git-crypt-setup-'));
  t.after(() => rm(folder, { recursive: true, force: true }));
  return folder;
}
